### PR TITLE
Implementation of bytearray and bincompare

### PR DESCRIPTION
**Pull request by helviojunior**
As the github merge all commits in only one push, i will comment all above:

Implementation of bytearray and bincompare

### Description/Motivation/Screenshots ###
This patch implement the Bincompare and Bytearray command wich will do the comparason of an Binary file and the memory position in order to find by badchars

### How Has This Been Tested? ###

| Architecture |          Yes/No          | Comments                                  |
| ------------ | :----------------------: | ----------------------------------------- |
| x86-32       | :heavy_check_mark: | Replace with :heavy_check_mark: if tested |
| x86-64       | :heavy_check_mark: |                                           |
| ARM          | :heavy_multiplication_x: |                                           |
| AARCH64      | :heavy_multiplication_x: |                                           |
| MIPS         | :heavy_multiplication_x: |                                           |
| POWERPC      | :heavy_multiplication_x: |                                           |
| SPARC        | :heavy_multiplication_x: |                                           |
| RISC-V       | :heavy_multiplication_x: |                                           |
| `make tests` | :heavy_check_mark: |                                           |

### Checklist ###

- [X] My PR was done against the `dev` branch, not `master`.
- [X] My code follows the code style of this project.
- [X] My change includes a change to the documentation, if required.
- [X] My change adds tests as appropriate.
- [X] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

File: docs/commands/bincompare.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+## Command bincompare
+
+The `bincompare` command will compare a binary file with memory data in order to find diferences between file and memory data. This command is primary used to check badchars in a exploitation process.
+
+`bincompare` require options:
+
+   * `-f` (for `file`) requires the full path of binary file to be compared.
```
Comment (Grazfather) [inline docs/commands/bincompare.md:7]:
```suggestion
* `-f` (for `file`) - the full path of binary file to be compared.
```

File: docs/commands/bincompare.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+   * `-a` (for `address`) requires the memory address to be compared with file data
```
Comment (Grazfather) [inline docs/commands/bincompare.md:8]:
```suggestion
* `-a` (for `address`) - the memory address to be compared with the file data.
```

File: docs/commands/bincompare.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ f0 |f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff| file
+    |                                               | memory
+    +-----------------------------------------------+
+
+[+] Badchars found: 05, 1f, 3f
+```
+
+Example with badchars and truncated buffer:
+```
+gef>  bincompare -f bytearray.bin -a 0x56557008
```
Comment (Grazfather) [inline docs/commands/bincompare.md:98]: are you using python2? Why is this the python2 gef prompt?

File: docs/commands/bytearray.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+## Command bytearray
+
+The `bytearray` generate a binary with data between 0x01 and 0xff. In general the created file is used to compare (using the `bincompare` command) with a memory data in order to check badchars.
```
Comment (Grazfather) [inline docs/commands/bytearray.md:3]:
```suggestion
The `bytearray` command generates a binary file with data between 0x01 and 0xff. In general the created file is used to compare (using the `bincompare` command) with a memory data in order to check badchars.
```

What are badchars?


can you link to bincompare's docs?

File: docs/commands/bytearray.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f"
+"\x20\x21\x22\x23\x24\x25\x26\x27\x28\x29\x2a\x2b\x2c\x2d\x2e\x2f\x30\x31\x32\x33\x34\x35\x36\x37\x38\x39\x3a\x3b\x3c\x3d\x3e\x3f"
+"\x40\x41\x42\x43\x44\x45\x46\x47\x48\x49\x4a\x4b\x4c\x4d\x4e\x4f\x50\x51\x52\x53\x54\x55\x56\x57\x58\x59\x5a\x5b\x5c\x5d\x5e\x5f"
+"\x60\x61\x62\x63\x64\x65\x66\x67\x68\x69\x6a\x6b\x6c\x6d\x6e\x6f\x70\x71\x72\x73\x74\x75\x76\x77\x78\x79\x7a\x7b\x7c\x7d\x7e\x7f"
+"\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f"
+"\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf"
+"\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf"
+"\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xfc\xfd\xfe\xff"
+
+[+] Done, wrote 256 bytes to file bytearray.txt
```
Comment (Grazfather) [inline docs/commands/bytearray.md:23]: we should be able to specify where the file is written.

File: gef.py [hunk 6]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+            elif o == "-h":
+                self.usage()
+                return
+
+        if not filename or not start_addr:
+            err("No file and/or address specified")
+            return
+
+        if not os.path.isfile(filename):
+            err("Especified file '{:s}' not exists".format(filename))
```
Comment (Grazfather) [inline gef.py:9223]:
```suggestion
            err("Specified file '{:s}' does not exist".format(filename))
```

File: gef.py [hunk 7]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+            return
+
+        f=open(filename, "rb")
+        file_data = f.read()
+        f.close()
```
Comment (Grazfather) [inline gef.py:9228]:
```suggestion
    with open(filename, "rb") as f:
        file_data = f.read()
```

File: gef.py [hunk 8]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        size = len(file_data)
+
+        if size < 8:
+            err("Error - file does not contain enough bytes (min 8 bytes needed)")
+            return
+
+        try:
+            memory_data = read_memory(start_addr, size)
+        except gdb.MemoryError:
+            err("Cannot reach memory {:#x}".format(start_addr))
```
Comment (Grazfather) [inline gef.py:9239]:
```suggestion
            err("Cannot read memory {:#x}".format(start_addr))
```

Can we more specifically show which addresses were invalid?

File: gef.py [hunk 9]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+            return
+
+
+        result_table = []
+        badchars = ""
+        cnt = 0
+        corrupted = -1
```
Comment (Grazfather) [inline gef.py:9246]: Use `None`

File: gef.py [hunk 10]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+
+    def print_line(self, line, data, label):
+        l = []
+        for d in data:
+            l.append(d)
+        r = 16 - len(l)
+        for i in range(0,r):
+            l.append("--")
+
+        gef_print(" {:s} |{:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s}| {:s}".format(line, l[0], l[1], l[2], l[3], l[4], l[5], l[6], l[7], l[8], l[9], l[10], l[11], l[12], l[13], l[14], l[15], label))
```
Comment (Grazfather) [inline gef.py:9301]: there are a lot nicer ways to do this.

File: gef.py [hunk 11]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+
+@register_command
+class BytearrayCommand(GenericCommand):
+    """BytearrayCommand: Generate a bytearray to be compared with possible badchars.
+Function ported from mona.py"""
```
Comment (Grazfather) [inline gef.py:9306]: what is mona.py?

File: gef.py [hunk 12]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        arrayfile = "bytearray.txt"
+        binfile = open(binfilename,"wb")
+        binfile.write(binarray)
+        binfile.close()
+
+        txtfile = open(arrayfile,"w+")
+        txtfile.write(output)
+        txtfile.close()
+
+        info("Done, wrote %d bytes to file %s" % (len(arraytable),arrayfile))
```
Comment (Grazfather) [inline gef.py:9419]: .format

File: gef.py [hunk 13]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        info("Binary output saved in %s" % binfilename)
+
+        return
+
+    def hex2bin(self, pattern):
+        """
+        Converts a hex string (\\x??\\x??\\x??\\x??) to real hex bytes
```
Comment (Grazfather) [inline gef.py:9426]: why do we need to save the file hex escaped?

File: gef.py [hunk 14]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        pattern = pattern.replace("\\x", "")
+        pattern = pattern.replace("\"", "")
+        pattern = pattern.replace("\'", "")
+        return binascii.unhexlify(pattern)
+        #return ''.join([binascii.unhexlify(i+j) for i,j in zip(pattern[0::2],pattern[1::2])])
+
+    def cleanHex(self, hex):
+
+        return ''.join(filter(self.permited_char, hex))
+        #return hex
```
Comment (Grazfather) [inline gef.py:9443]: x

File: gef.py [hunk 15]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        """
+        pattern = pattern.replace("\\x", "")
+        pattern = pattern.replace("\"", "")
+        pattern = pattern.replace("\'", "")
+        return binascii.unhexlify(pattern)
+        #return ''.join([binascii.unhexlify(i+j) for i,j in zip(pattern[0::2],pattern[1::2])])
+
+    def cleanHex(self, hex):
+
+        return ''.join(filter(self.permited_char, hex))
```
Comment (Grazfather) [inline gef.py:9442]: ""

File: gef.py [hunk 16]
```diff
@@ -9181,6 +9181,275 @@ def get_settings_path(self):
...
+        #return hex
+
+    def hex2int(self, hex):
+        return int(hex,16)
+
+    def permited_char(self, s):
```
Comment (Grazfather) [inline gef.py:9448]: permited -> permitted

File: gef.py (hunk 14 above)
Comment (helviojunior) [reply gef.py:9443]: i will commit changes

File: gef.py (hunk 15 above)
Comment (helviojunior) [reply gef.py:9442]: i will commit changes

File: gef.py (hunk 16 above)
Comment (helviojunior) [reply gef.py:9448]: I will commit

File: docs/commands/bincompare.md (hunk 3 above)
Comment (helviojunior) [reply docs/commands/bincompare.md]: Changed.

File: docs/commands/bincompare.md [hunk 17]
```diff
@@ -0,0 +1,137 @@
+## Command bincompare
+
+The `bincompare` command will compare a binary file with memory data in order to find diferences between file and memory data. This command is primary used to check badchars in a exploitation process.
+
+`bincompare` require options:
```
[the remark this answers is not in the document]
Comment (helviojunior) [reply docs/commands/bincompare.md]: Commited

**Comment by hugsy** [comment]
Please move this PR into [`gef-extras/scripts`](https://github.com/hugsy/gef-extras)

File: gef.py [hunk 18]
```diff
@@ -9181,6 +9181,125 @@ def get_settings_path(self):
         path = os.path.realpath(path)
         return path if os.path.isdir(path) else None
 
+@register_command
+class BincompareCommand(GenericCommand):
+    """BincompareCommand: compare an binary file with the memory position looking for badchars."""
+    _cmdline_ = "bincompare"
+    _syntax_  = "{:s} -f FILE -a MEMORY_ADDRESS [-h]".format(_cmdline_)
```
Comment (hugsy) [inline gef.py:9189]: Missing `_example_`, would be useful to understand the command.

**Comment by helviojunior** [comment]
> Please move this PR into [`gef-extras/scripts`](https://github.com/hugsy/gef-extras)

Ok, i will close this pull request, move to extras and create a new one.

File: gef.py [hunk 19]
```diff
@@ -9181,6 +9181,125 @@ def get_settings_path(self):
...
+        start_addr = None
+        size = 0
+        file_data = None
+        memory_data = None
+
+        opts, args = getopt.getopt(argv, "f:a:ch")
+        for o,a in opts:
+            if   o == "-f": filename = a
+            elif o == "-a": start_addr = int(gdb.parse_and_eval(a))
+            elif o == "-h":
```
Comment (hugsy) [inline gef.py:9215]: `-h` is unnecessary, the docstring of the command and the `usage` build the help menu if you call the GDB `help <command>`

**Comment by helviojunior** [comment]
Closed to move to gef extras.

**Pull request closed without merging**